### DISCUSSION TITLE
check buffer status before writing  to decoder .

### DISCUSSION THREAD

**Discussion by aqildad-create** [pschatzmann/arduino-audio-tools#761] [Q&A] [answered]
I am writing to decoder and it is working perfectly  but i want to find first the buffer status then write because i don't want to write fast or overflow the buffer . how to do like that ? 

EncodedAudioStream decStream(i2s, dec);
.
.
decStream.write((uint8_t *)Filedatas.charF , Filedatas.FiLEDATAsze);  ------this work perfect but need to now the buffer status first then write .

**Comment by pschatzmann** [comment, marked as answer]
In your code snipplet there is no buffer that can overflow. 
I2SStream is using a blocking write!
However it is good practice to check the result of any write() call

**Reply by aqildad-create** [reply]
can we check the status of buffer of I2SStream  that how much it is filled at any point of time ? 

before write again to decoder I want to check the i2sbuffer ? 

I2SStream is using a blocking write but decoder.write is not blocking write because i used decStream.write((uint8_t *)Filedatas.charF , Filedatas.FiLEDATAsze); ------ within OnDataRecv event of ESPNOW and it work successfully  with no data loss .or very minimum data loss . what is your suggestion for this ?  If decoder.write zIS BLOCKNG WRTE THEN WHY i AM NOT GETTNG ANY LOSS IN DATA RECEAVE ? 

but i am worry if espnow data receive is fast or slow in both cases buffer of I2SStream  may be over flow(or data lost ) or if espnow data receive is slow then I2SStream   will empty . so that why  want to check the buffer status of I2SStream   . 

Is this possible ? If yes how ? or if no then how to know the status buffer  of I2SStream   ?

actually right now i am reading mp3 form sd card of esp1  and via espnow  and i am sending to esp2 were  use decoder and play . so i want to know how fast should i read data from file .

**Comment by pschatzmann** [comment]
In general you can call the standard Arduino method availableForWrite(). But in the case of I2S this will always provide DEFAULT_BUFFER_SIZE because there is no functionality to to check the internal buffers!

**Reply by aqildad-create** [reply]
So how to control the writes ? at how much speed we should write to decoder  ?

**Reply by pschatzmann** [reply]
I have already answered this question: It is a blocking write! 
It waits until it can process the request and you don't need to do anything...

Should I ever decide to change this you will need to check for the processed bytes and just resend the unprocessed ones.
This functionality is built into the StreamCopy class and you don't need to reinvent the wheel again.

**Reply by aqildad-create** [reply]
I cant use StreamCopy because mp3 file is  in esp1 ,  i have to read 240 bytes from file wth some other bytes to esp2 va espnow were decoder is place .

**Reply by pschatzmann** [reply]
Yes, that's because you ignored my advice to write your own container class!

**Comment by aqildad-create** [comment]
Reading from i2s then encodersbc  then espnow then decoder SBC and i2s .
this work . I thing it was sink automatically because of i2s at both end
starting and ending .

While reading from SD then espnow and decoder mp3 and then i2s out . this
do not work may be because reading from SD is fast then writing to i2s at
esp32 .

Am I correct ?

I notice a file of approx 700kbytes can send in 2mint via espnow and it
take 6 minutes to play . . . either I need to send slow the data in order
to sync it .. For 700kbyes of mp3 file should send in 6mint and parallel
play in same 6 minutes right ?

**Reply by pschatzmann** [reply]
In esp-now the receiver is sending confirmations and the i2s output is blocking, so it should automatically throttle to the correct speed

**Reply by aqildad-create** [reply]
yes , i send acknowledgment like below  its working and playing but their are sound jerks which is because delay after decstream write and new data coming may be . how to overcome this ?

OnDataRecv(
 .
.
decStreammp3.write((uint8_t *)Filedatas.charF , Filedatas.FiLEDATAsze);
esp_now_send(broadcastAddres, (uint8_t *) &outgoingSetpoints, sizeof(outgoingSetpoints)); //acnowledgment send 

}

**Comment by pschatzmann** [comment]
Just wanted to let know know that I finally manage to test my custom container class: [for details see here](https://www.pschatzmann.ch/home/2023/04/28/arduino-audio-tools-codecs-and-containers/ ). You can write you own additional data in the encoder with writeMeta (const uint8_t *data, size_t len) and retrieve it in the decoder with a callback method.
